### PR TITLE
feat(api): `nvim_get_chan_info` pass 0 for current channel

### LINKED ISSUE #21291: nvim_get_chan_id, or return channel id from nvim_set_client_info
### Feature already in Vim?

N/A

### Feature description

Right now you can call `nvim_get_api_info` to get the channel ID and the API metadata, but often you want these pieces of information at different times - the channel is relevant when you're running the plugin, while the API metadata is mostly relevant for generating API bindings. The API metadata still may be relevant at runtime for validation purposes, but libraries that choose not to do this validation (instead letting invalid calls result in error responses from Neovim) only need the channel ID.

### PR DESCRIPTION
**Pull request by nikolightsaber**
closes #21291

Hi first time doing this. I tried to follow the guidelines. I hope documentation and testing is good enough.

For testing I tried to follow `nvim_get_api_info()` but i did not find unit tests on this. I figured changing some of the `nvim_get_api_info()[1]` should be good enough for such a simple change.

Thanks for the feedback.

### CODE REVIEW

**Review by justinmk — CHANGES_REQUESTED** [review]
Let's try to avoid adding a new api function. Why can't we do this: https://github.com/neovim/neovim/issues/21291#issuecomment-1455212995

> Instead, we could make `nvim_get_chan_info(0)` (that is, "zero" means "current channel") work.

If that doesn't work, I'm starting to think we should relax `:help api-contract` to allow `void` return values to be changed.

**Comment by nikolightsaber** [comment]
@justinmk I think i succeeded in `nvim_get_chan_info` handling 0 as an argument.

I figured that channel 0 is impossible as nvim always seem to start at 3.

I'm a bit confused on the doc and linting pipeline failing. Could it be an issue unrelated to my changes ?

The test that fail seem to be my changes as the have to to with channels. I will have a look at them if needed.

I would appreciate a review at this stage.

Thanks

File: runtime/doc/news.txt [hunk 1]
```diff
@@ -405,6 +405,9 @@ The following changes to existing APIs or features add new behavior.
 • Attempting to set an invalid keycode option (e.g. `set t_foo=123`) no longer
   gives an error.
 
+• Allow to pass 0 to |nvim_get_chan_info| to get channel info for current
+  channel
```
Comment (zeertzjq) [inline runtime/doc/news.txt:409]:
```suggestion
• Passing 0 to |nvim_get_chan_info| gets channel info for the current channel.
```

File: src/nvim/api/vim.c [hunk 2]
```diff
@@ -1633,12 +1634,16 @@ void nvim_set_client_info(uint64_t channel_id, String name, Dictionary version,
...
-Dictionary nvim_get_chan_info(Integer chan, Error *err)
+Dictionary nvim_get_chan_info(uint64_t channel_id, Integer chan, Error *err)
   FUNC_API_SINCE(4)
 {
   if (chan < 0) {
     return (Dictionary)ARRAY_DICT_INIT;
   }
+  if (chan == 0) {
+    assert(channel_id <= INT64_MAX);
+    chan = (Integer) channel_id;
```
Comment (zeertzjq) [inline src/nvim/api/vim.c:1645]:
```suggestion
    chan = (Integer)channel_id;
```

File: src/nvim/api/vim.c [hunk 3]
```diff
@@ -1633,12 +1634,16 @@ void nvim_set_client_info(uint64_t channel_id, String name, Dictionary version,
...
 ///
-Dictionary nvim_get_chan_info(Integer chan, Error *err)
+Dictionary nvim_get_chan_info(uint64_t channel_id, Integer chan, Error *err)
   FUNC_API_SINCE(4)
 {
   if (chan < 0) {
     return (Dictionary)ARRAY_DICT_INIT;
   }
+  if (chan == 0) {
+    assert(channel_id <= INT64_MAX);
```
Comment (zeertzjq) [inline src/nvim/api/vim.c:1644]: Internal calls will fail this assertion:
```c
/// Mask for all internal calls
#define INTERNAL_CALL_MASK (((uint64_t)1) << (sizeof(uint64_t) * 8 - 1))

/// Internal call from Vimscript code
#define VIML_INTERNAL_CALL INTERNAL_CALL_MASK

/// Internal call from Lua code
#define LUA_INTERNAL_CALL (VIML_INTERNAL_CALL + 1)
```

File: src/nvim/api/vim.c [hunk 4]
```diff
@@ -1633,12 +1634,16 @@ void nvim_set_client_info(uint64_t channel_id, String name, Dictionary version,
...
 ///                 |nvim_set_client_info()|.
 ///
-Dictionary nvim_get_chan_info(Integer chan, Error *err)
+Dictionary nvim_get_chan_info(uint64_t channel_id, Integer chan, Error *err)
   FUNC_API_SINCE(4)
 {
   if (chan < 0) {
     return (Dictionary)ARRAY_DICT_INIT;
   }
+  if (chan == 0) {
```
Comment (zeertzjq) [inline src/nvim/api/vim.c:1643]:
```suggestion
  if (chan == 0 && !is_internal_call(channid_id)) {
```

**Comment by nikolightsaber** [comment]
Thanks for the review @zeertzjq I implemented the changes.

File: test/functional/api/vim_spec.lua [hunk 5]
```diff
@@ -2449,7 +2449,8 @@ describe('API', function()
     }
 
     it('returns {} for invalid channel', function()
-      eq({}, api.nvim_get_chan_info(0))
+      -- 0 should return current channel
+      eq(testinfo, api.nvim_get_chan_info(0))
```
Comment (zeertzjq) [inline test/functional/api/vim_spec.lua:2453]: I think this should be moved to the `stream=stdio channel` block, before both `eq(testinfo, api.nvim_get_chan_info(1))` assertions.

**Review by justinmk — APPROVED** [review]
please explain what problem this is solving, in the commit message

**Review by zeertzjq — APPROVED** [review]

**Comment by nikolightsaber** [comment]
Hi @justinmk, @zeertzjq thanks for the help with this.  I enjoyed learning about the api. Quite clever with the generated code. 

I hope I can contribute some more if I find some time.

**Pull request merged**